presto-native-execution/presto_cpp/main/SystemConnector.cpp

```diff
@@ -0,0 +1,297 @@
+  auto systemTableHandle =
+      std::dynamic_pointer_cast<SystemTableHandle>(tableHandle);
+  VELOX_CHECK_NOT_NULL(
+      systemTableHandle,
+      "TableHandle must be an instance of SystemTableHandle");
+  VELOX_USER_CHECK_EQ(
+      systemTableHandle->schemaName(),
+      "runtime",
+      "SystemConnector supports only runtime schema");
+  VELOX_USER_CHECK_EQ(
+      systemTableHandle->tableName(),
+      "tasks",
+      "SystemConnector supports only tasks table");
+
+  taskTableHandle_ = systemTableHandle;
+  outputColumnMappings_.reserve(outputType->names().size());
+  auto taskSchema = taskTableHandle_->taskSchema();
+  for (const auto& outputName : outputType->names()) {
+    auto it = columnHandles.find(outputName);
+    VELOX_CHECK(
+        it != columnHandles.end(),
+        "ColumnHandle is missing for output column '{}'",
+        outputName);
+
+    auto handle = std::dynamic_pointer_cast<SystemColumnHandle>(it->second);
+    VELOX_CHECK_NOT_NULL(
+        handle,
+        "ColumnHandle must be an instance of SystemColumnHandle "
+        "for '{}' on table '{}'",
+        handle->name());
+
+    auto columnIndex = taskSchema->getChildIdxIfExists(handle->name());
+    VELOX_CHECK(columnIndex.has_value());
+    outputColumnMappings_.push_back(columnIndex.value());
+  }
+
+  outputType_ = outputType;
+  taskTableResult_ = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(taskSchema, 0, pool_));
+}
+
+void SystemDataSource::addSplit(
+    std::shared_ptr<connector::ConnectorSplit> split) {
+  VELOX_CHECK_NULL(
+      currentSplit_,
+      "Previous split has not been processed yet. Call next() to process the split.");
+  currentSplit_ = std::dynamic_pointer_cast<SystemSplit>(split);
+  VELOX_CHECK(currentSplit_, "Wrong type of split for SystemDataSource.");
+  VELOX_USER_CHECK_EQ(
+      currentSplit_->schemaName(),
+      "runtime",
+      "SystemConnector supports only runtime schema");
+  VELOX_USER_CHECK_EQ(
+      currentSplit_->tableName(),
+      "tasks",
+      "SystemConnector supports only tasks table");
+}
+
+namespace {
+void getTaskResults(
+    const TaskMap& taskMap,
+    const RowVectorPtr& taskTableResult) {
+  auto numRows = taskMap.size();
+  taskTableResult->resize(numRows);
+  for (auto i = 0; i < taskTableResult->type()->size(); i++) {
+    taskTableResult->childAt(i)->resize(numRows);
+  }
+
+  auto toMillis = [](int64_t nanos) -> int64_t { return nanos / 1'000'000; };
+  int i = 0;
+  for (const auto& taskEntry : taskMap) {
+    auto task = taskEntry.second;
+    auto taskInfo = task->updateInfo();
+    taskTableResult->childAt(0)->asFlatVector<StringView>()->set(
+        i, StringView(taskInfo.nodeId));
+    taskTableResult->childAt(1)->asFlatVector<StringView>()->set(
+        i, StringView(taskInfo.taskId));
+    std::string temp = fmt::format("{}", task->id.stageExecutionId());
+    taskTableResult->childAt(2)->asFlatVector<StringView>()->set(
```

Comment:
This code is hard to read because it is not clear what column 2 or 5 means. Perhaps, create an enum.